### DISCUSSION TITLE
How do I add the nonce attribute to a dynamically inserted script?

### DISCUSSION THREAD

**Discussion by GavinLam164** [vercel/next.js#50653] [App Router]
I am trying to set the nonce attribute to all dynamically inserted scripts, as I am going to integrate the CSP policy into the site.
Is there a better way? Thanks for your reply.
https://github.com/vercel/next.js/blob/eb85e79941ca77cd5b111d247718c9f3da3fa5cb/packages/next/src/server/app-render/app-render.tsx#LL1238C1-L1243C6

**Comment by leerob** [comment]
Hey folks, wanted to swing back here with an update. After digging through many different issues and discussions, I've [**made a new page in the documentation**](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy) ([PR](https://github.com/vercel/next.js/pull/54601)) specifically for Content Security Policy and nonces. This docs page: 

- Explains how to generate a `nonce` with Middleware
- Shows how to consume the `nonce` in a route with `headers()`
- Shows a complete CSP without needing to use any `unsafe`
- Shows how to ignore the `nonce` Middleware from running on prefetches / static assets

Further, we've patched some bugs and made improvements to `nonce` handling in Next.js that will be available in the latest `canary` version (for those of you time traveling from the future, upgrade to Next.js 13.5). We also updated the `with-strict-csp` example in the `examples/` folder, which is backlinked from the [new documentation page](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy).

Really hope this helps out, thank you all 🙏 I'll be closing this discussion out. To continue the discussion, [please go here](https://github.com/vercel/next.js/discussions/54907).
